File: apps/desktop/src-tauri/examples/seed_dev_db.rs
//! Development fixture seeder for Singapur Cards.
//!
//! Populates a local SQLite database with realistic dictionaries, cards, collections,
//! memberships, and review events for local development and smoke testing.
//!
//! Usage:
//!   cargo run --example seed_dev_db --manifest-path src-tauri/Cargo.toml
//!   cargo run --example seed_dev_db --manifest-path src-tauri/Cargo.toml -- --seed 42
//!   cargo run --example seed_dev_db --manifest-path src-tauri/Cargo.toml -- --db /tmp/dev.db

use fake::faker::lorem::en::{Sentence, Word};
use fake::Fake;
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use rusqlite::Connection;
use singapur_cards_lib::db::queries;
use singapur_cards_lib::db::schema::run_migrations;
use singapur_cards_lib::models::*;
use std::env;
use std::path::PathBuf;

const CARD_COUNT: usize = 30;
const COLLECTION_COUNT: usize = 4;

fn parse_args() -> (u64, PathBuf) {
    let args: Vec<String> = env::args().collect();
    let mut seed: u64 = 12345;
    let mut db_path = PathBuf::from("/tmp/singapur_cards_dev.db");

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--seed" => {
                i += 1;
                seed = args.get(i).and_then(|s| s.parse().ok()).unwrap_or(seed);
            }
            "--db" => {
                i += 1;
                if let Some(p) = args.get(i) {
                    db_path = PathBuf::from(p);
                }
            }
            _ => {}
        }
        i += 1;
    }
    (seed, db_path)
}

fn main() -> rusqlite::Result<()> {
    let (seed, db_path) = parse_args();
    println!("Seeding database at: {} (seed={})", db_path.display(), seed);

    if db_path.exists() {
        std::fs::remove_file(&db_path).ok();
    }

    let conn = Connection::open(&db_path)?;
    run_migrations(&conn).expect("Failed to run migrations");

    let mut rng = StdRng::seed_from_u64(seed);

    // ── Dictionaries ──────────────────────────────────────────────────────────
    let lang_pairs = [("en", "de"), ("en", "es"), ("de", "en")];
    let mut all_entry_ids: Vec<(String, String)> = vec![]; // (entry_id, language_from)

    for (dict_idx, (lang_from, lang_to)) in lang_pairs.iter().enumerate() {
        let dict_id = format!("dict-{dict_idx}");
        let dict = Dictionary {
            id: dict_id.clone(),
            name: format!("{} → {} Dictionary", lang_from.to_uppercase(), lang_to.to_uppercase()),
            language_from: lang_from.to_string(),
            language_to: lang_to.to_string(),
            source_filename: format!("{lang_from}_{lang_to}.dsl"),
            source_path: None,
            import_status: "ready".to_string(),
            entry_count: 0,
            last_error: None,
            created_at: "2026-01-01T00:00:00Z".to_string(),
            updated_at: "2026-01-01T00:00:00Z".to_string(),
        };
        queries::insert_dictionary(&conn, &dict)?;

        let entry_count = 20usize;
        for j in 0..entry_count {
            let headword: String = Word().fake_with_rng(&mut rng);
            let definition: String = Sentence(3..8).fake_with_rng(&mut rng);
            let example: String = Sentence(5..12).fake_with_rng(&mut rng);
            let entry_id = format!("entry-{dict_idx}-{j}");

            let entry = DictionaryEntry {
                id: entry_id.clone(),
                dictionary_id: dict_id.clone(),
                headword: headword.clone(),
                normalized_headword: headword.to_lowercase(),
                transcription: None,
                definition_text: definition,
                example_text: Some(example),
                audio_reference: None,
                source_order: j as i64,
                created_at: "2026-01-01T00:00:00Z".to_string(),
            };
            queries::insert_dictionary_entry(&conn, &entry)?;
            all_entry_ids.push((entry_id, lang_from.to_string()));
        }

        conn.execute(
            "UPDATE dictionaries SET entry_count=? WHERE id=?",
            rusqlite::params![entry_count as i64, dict_id],
        )?;
    }

    // ── Collections ───────────────────────────────────────────────────────────
    let collection_names = ["Core Vocabulary", "Verbs", "Nouns", "Daily Use"];
    let collection_ids: Vec<String> = collection_names[..COLLECTION_COUNT]
        .iter()
        .enumerate()
        .map(|(i, name)| {
            let coll = Collection {
                id: format!("coll-{i}"),
                name: name.to_string(),
                description: Some(format!("A curated set of {}", name.to_lowercase())),
                created_at: "2026-01-01T00:00:00Z".to_string(),
                updated_at: "2026-01-01T00:00:00Z".to_string(),
            };
            queries::insert_collection(&conn, &coll).expect("insert collection");
            coll.id
        })
        .collect();

    // ── Cards ─────────────────────────────────────────────────────────────────
    let statuses = ["unreviewed", "not_learned", "learned"];
    let mut card_ids: Vec<String> = vec![];

    for i in 0..CARD_COUNT {
        let headword: String = Word().fake_with_rng(&mut rng);
        let answer: String = Sentence(2..6).fake_with_rng(&mut rng);
        let example: String = Sentence(5..12).fake_with_rng(&mut rng);
        let status = statuses[i % statuses.len()];
        let card_id = format!("card-{i}");
        let lang = if i % 3 == 2 { "de" } else { "en" };

        let card = Card {
            id: card_id.clone(),
            language: lang.to_string(),
            headword,
            answer_text: answer,
            example_text: Some(example),
            notes: None,
            audio_reference: None,
            source_entry_ids: if i < all_entry_ids.len() {
                vec![all_entry_ids[i].0.clone()]
            } else {
                vec![]
            },
            learning_status: status.to_string(),
            created_at: format!("2026-01-{:02}T00:00:00Z", (i % 28) + 1),
            updated_at: format!("2026-01-{:02}T00:00:00Z", (i % 28) + 1),
            last_reviewed_at: if status != "unreviewed" {
                Some("2026-03-01T00:00:00Z".to_string())
            } else {
                None
            },
        };

        if let Err(e) = queries::insert_card(&conn, &card) {
            eprintln!("  Skipping card-{i} ({}): {e}", card.headword);
            continue;
        }
        card_ids.push(card_id);
    }

    // ── Memberships ───────────────────────────────────────────────────────────
    for (i, card_id) in card_ids.iter().enumerate() {
        let num_colls: usize = rng.gen_range(0..=2);
        if num_colls == 0 {
            continue;
        }
        let chosen: Vec<String> = {
            let mut idxs: Vec<usize> = (0..collection_ids.len()).collect();
            idxs.sort_by_key(|_| rng.gen::<u64>());
            idxs[..num_colls.min(collection_ids.len())]
                .iter()
                .map(|&idx| collection_ids[idx].clone())
                .collect()
        };
        queries::replace_card_memberships(&conn, card_id, &chosen, "2026-01-01T00:00:00Z")
            .expect("insert memberships");
        let _ = i; // suppress warning
    }

    // ── Review events ─────────────────────────────────────────────────────────
    let review_results = ["learned", "not_learned"];
    let mut event_idx = 0usize;
    for card_id in card_ids.iter().filter(|_| rng.gen_bool(0.6)) {
        let result = review_results[event_idx % review_results.len()];
        let event = ReviewEvent {
            id: format!("review-{event_idx}"),
            card_id: card_id.clone(),
            result: result.to_string(),
            reviewed_at: "2026-03-01T00:00:00Z".to_string(),
        };
        queries::insert_review_event(&conn, &event).expect("insert review event");
        event_idx += 1;
    }

    println!("✓ Seeded {} dictionaries ({} entries each)", lang_pairs.len(), 20);
    println!("✓ Seeded {} collections", COLLECTION_COUNT);
    println!("✓ Seeded {} cards", card_ids.len());
    println!("✓ Seeded {} review events", event_idx);
    println!("\nDatabase ready at: {}", db_path.display());
    println!("Run with --seed <N> to reproduce this exact dataset.");

    Ok(())
}

File: apps/mobile/scripts/seed-dev-db.ts
/**
 * Development fixture seeder for Singapur Cards (mobile).
 *
 * Populates a local SQLite database with realistic dictionaries, cards,
 * collections, memberships, and review events for local development.
 *
 * Usage:
 *   npm run db:seed
 *   npm run db:seed -- --db /tmp/custom.db
 *   npm run db:seed -- --seed 42
 */

import Database from 'better-sqlite3';
import { drizzle } from 'drizzle-orm/better-sqlite3';
import { faker } from '@faker-js/faker';
import { readFileSync, rmSync, existsSync } from 'fs';
import { join } from 'path';
import * as schema from '../src/db/schema';
import {
  aiCredentials,
  cards,
  chatConversations,
  chatMessages,
  collectionMemberships,
  collections,
  customChatModels,
  dictionaries,
  dictionaryEntries,
  languages,
  reviewEvents,
} from '../src/db/schema';

// ── Config ────────────────────────────────────────────────────────────────────

const CARD_COUNT = 30;
const COLLECTION_COUNT = 4;
const ENTRIES_PER_DICT = 20;
const LANG_PAIRS: [string, string][] = [
  ['en', 'de'],
  ['en', 'es'],
  ['de', 'en'],
];
const COLLECTION_NAMES = ['Core Vocabulary', 'Verbs', 'Nouns', 'Daily Use'];

// ── Arg parsing ───────────────────────────────────────────────────────────────

function parseArgs(): { seed: number; dbPath: string } {
  const args = process.argv.slice(2);
  let seed = 12345;
  let dbPath = '/tmp/singapur_mobile_dev.db';

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--seed' && args[i + 1]) {
      seed = parseInt(args[++i], 10);
    } else if (args[i] === '--db' && args[i + 1]) {
      dbPath = args[++i];
    }
  }
  return { seed, dbPath };
}

// ── Schema bootstrap ──────────────────────────────────────────────────────────

function applySchema(sqlite: Database.Database): void {
  sqlite.pragma('journal_mode = WAL');
  sqlite.pragma('foreign_keys = ON');

  const migrationPath = join(
    __dirname,
    '../src/db/migrations/0000_gigantic_sharon_ventura.sql',
  );
  const sql = readFileSync(migrationPath, 'utf8');

  // Drizzle expo migrations use `--> statement-breakpoint` as a delimiter.
  const statements = sql
    .split('--> statement-breakpoint')
    .map((s) => s.trim())
    .filter(Boolean);

  for (const stmt of statements) {
    sqlite.exec(stmt);
  }
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  const { seed, dbPath } = parseArgs();
  console.log(`Seeding database at: ${dbPath} (seed=${seed})`);

  if (existsSync(dbPath)) rmSync(dbPath);

  const sqlite = new Database(dbPath);
  applySchema(sqlite);

  const db = drizzle(sqlite, { schema });

  faker.seed(seed);

  // ── Languages seed ─────────────────────────────────────────────────────────
  db.insert(languages)
    .values({ code: 'en', title: 'English', createdAt: now() })
    .onConflictDoNothing()
    .run();
  db.insert(languages)
    .values({ code: 'de', title: 'German', createdAt: now() })
    .onConflictDoNothing()
    .run();
  db.insert(languages)
    .values({ code: 'es', title: 'Spanish', createdAt: now() })
    .onConflictDoNothing()
    .run();

  // ── Dictionaries & entries ─────────────────────────────────────────────────
  const allEntries: { id: string; lang: string }[] = [];

  for (let dictIdx = 0; dictIdx < LANG_PAIRS.length; dictIdx++) {
    const [langFrom, langTo] = LANG_PAIRS[dictIdx];
    const dictId = `dict-${dictIdx}`;

    db.insert(dictionaries)
      .values({
        id: dictId,
        name: `${langFrom.toUpperCase()} → ${langTo.toUpperCase()} Dictionary`,
        languageFrom: langFrom,
        languageTo: langTo,
        sourceFilename: `${langFrom}_${langTo}.dsl`,
        importStatus: 'ready',
        entryCount: ENTRIES_PER_DICT,
        createdAt: now(),
        updatedAt: now(),
      })
      .run();

    for (let j = 0; j < ENTRIES_PER_DICT; j++) {
      const headword = faker.word.noun();
      const entryId = `entry-${dictIdx}-${j}`;

      db.insert(dictionaryEntries)
        .values({
          id: entryId,
          dictionaryId: dictId,
          headword,
          normalizedHeadword: headword.toLowerCase(),
          definitionText: faker.lorem.sentence({ min: 3, max: 7 }),
          exampleText: faker.lorem.sentence({ min: 5, max: 12 }),
          sourceOrder: j,
          createdAt: now(),
        })
        .run();

      allEntries.push({ id: entryId, lang: langFrom });
    }
  }

  // ── Collections ────────────────────────────────────────────────────────────
  const collectionIds: string[] = [];

  for (let i = 0; i < COLLECTION_COUNT; i++) {
    const collId = `coll-${i}`;
    const name = COLLECTION_NAMES[i];
    db.insert(collections)
      .values({
        id: collId,
        name,
        description: `A curated set of ${name.toLowerCase()}`,
        createdAt: now(),
        updatedAt: now(),
      })
      .run();
    collectionIds.push(collId);
  }

  // ── Cards ──────────────────────────────────────────────────────────────────
  const statuses = ['unreviewed', 'not_learned', 'learned'] as const;
  const cardIds: string[] = [];

  for (let i = 0; i < CARD_COUNT; i++) {
    const cardId = `card-${i}`;
    const status = statuses[i % statuses.length];
    const lang = i % 3 === 2 ? 'de' : 'en';
    const headword = faker.word.noun();
    const date = `2026-01-${String((i % 28) + 1).padStart(2, '0')}T00:00:00Z`;

    try {
      db.insert(cards)
        .values({
          id: cardId,
          language: lang,
          headword,
          answerText: faker.lorem.sentence({ min: 2, max: 6 }),
          exampleText: faker.lorem.sentence({ min: 5, max: 12 }),
          sourceEntryIds: JSON.stringify(
            i < allEntries.length ? [allEntries[i].id] : [],
          ),
          learningStatus: status,
          createdAt: date,
          updatedAt: date,
          lastReviewedAt: status !== 'unreviewed' ? '2026-03-01T00:00:00Z' : null,
        })
        .run();
      cardIds.push(cardId);
    } catch {
      console.warn(`  Skipping card-${i} (${headword}): duplicate headword+language`);
    }
  }

  // ── Memberships ────────────────────────────────────────────────────────────
  for (const cardId of cardIds) {
    const numColls = faker.number.int({ min: 0, max: 2 });
    if (numColls === 0) continue;

    const chosen = faker.helpers
      .shuffle([...collectionIds])
      .slice(0, numColls);

    for (const collectionId of chosen) {
      try {
        db.insert(collectionMemberships)
          .values({ collectionId, cardId, createdAt: now() })
          .onConflictDoNothing()
          .run();
      } catch {
        // ignore duplicate memberships
      }
    }
  }

  // ── Review events ──────────────────────────────────────────────────────────
  const reviewResults = ['learned', 'not_learned'] as const;
  let eventIdx = 0;

  for (const cardId of cardIds) {
    if (!faker.datatype.boolean(0.6)) continue;

    db.insert(reviewEvents)
      .values({
        id: `review-${eventIdx}`,
        cardId,
        result: reviewResults[eventIdx % reviewResults.length],
        reviewedAt: '2026-03-01T00:00:00Z',
      })
      .run();
    eventIdx++;
  }

  // ── AI credentials (one fixture entry) ────────────────────────────────────
  db.insert(aiCredentials)
    .values({
      id: 'cred-0',
      provider: 'openai',
      label: 'Dev key',
      isActive: true,
      createdAt: now(),
      updatedAt: now(),
    })
    .onConflictDoNothing()
    .run();

  // ── Chat fixtures ──────────────────────────────────────────────────────────
  db.insert(chatConversations)
    .values({
      id: 'conv-0',
      title: 'Vocabulary help',
      model: 'gpt-4o',
      createdAt: now(),
      updatedAt: now(),
    })
    .run();

  db.insert(chatMessages)
    .values([
      {
        id: 'msg-0',
        conversationId: 'conv-0',
        role: 'user',
        body: 'What does "Schadenfreude" mean?',
        createdAt: now(),
      },
      {
        id: 'msg-1',
        conversationId: 'conv-0',
        role: 'assistant',
        body: faker.lorem.paragraph(),
        createdAt: now(),
      },
    ])
    .run();

  // ── Custom models ──────────────────────────────────────────────────────────
  db.insert(customChatModels)
    .values({
      id: 'model-0',
      name: 'gpt-4o-mini',
      title: 'GPT-4o Mini',
      provider: 'openai',
      createdAt: now(),
    })
    .onConflictDoNothing()
    .run();

  sqlite.close();

  console.log(`✓ Seeded ${LANG_PAIRS.length} dictionaries (${ENTRIES_PER_DICT} entries each)`);
  console.log(`✓ Seeded ${COLLECTION_COUNT} collections`);
  console.log(`✓ Seeded ${cardIds.length} cards`);
  console.log(`✓ Seeded ${eventIdx} review events`);
  console.log(`\nDatabase ready at: ${dbPath}`);
  console.log('Run with --seed <N> to reproduce this exact dataset.');
}

function now(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
